File: JavaCode/src/com/ysx/leetcode/easy/TransposeMatrix.java
package com.ysx.leetcode.easy;

/**
 * @author youngbear
 * @email youngbear@aliyun.com
 * @date 2018/7/8 21:06
 * @blog https://blog.csdn.net/next_second
 * @github https://github.com/YoungBear
 * @description 矩阵的转置
 * Leetcode 868. Transpose Matrix
 * https://leetcode.com/problems/transpose-matrix/description/
 */
public class TransposeMatrix {

    public int[][] transpose(int[][] A) {
        int[][] B = new int[A[0].length][A.length];
        for (int i = 0; i < B.length; i++) {
            for (int j = 0; j < B[i].length; j++) {
                B[i][j] = A[j][i];
            }
        }
        return B;
    }

    public static void main(String[] args) {
        int[][] A = {{1,2,3},{4,5,6},{7,8,9}};
        printMatrix(A);
        System.out.println("----------------");
        TransposeMatrix test = new TransposeMatrix();
        int[][] B = test.transpose(A);
        printMatrix(B);

        int[][] A2 = {{1,2,3},{4,5,6}};
        printMatrix(A2);
        System.out.println("----------------");
        int[][] B2 = test.transpose(A2);
        printMatrix(B2);
    }

    public static void printMatrix(int[][] A){
        for (int i = 0; i< A.length; i++) {
            for (int j = 0; j < A[i].length; j++) {
                System.out.print(A[i][j] + " ");
            }
            System.out.println();
        }
    }
}

File: KotlinCode/src/leetcode/easy/TransposeMatrix.kt
package leetcode.easy

/**
 * @author youngbear
 * @email youngbear@aliyun.com
 * @date 2018/7/8 21:31
 * @blog https://blog.csdn.net/next_second
 * @github https://github.com/YoungBear
 * @description 矩阵的转置
 * Leetcode 868. Transpose Matrix
 * https://leetcode.com/problems/transpose-matrix/description/
 */
class TransposeMatrix {
    fun transpose(A: Array<IntArray>): Array<IntArray> {
        val B = Array(A[0].size) { IntArray(A.size) }
        for (i in B.indices) {
            for (j in 0 until B[i].size) {
                B[i][j] = A[j][i]
            }
        }
        return B
    }
    companion object {

        @JvmStatic
        fun main(args: Array<String>) {
            val test = TransposeMatrix()

            val A = arrayOf(intArrayOf(1, 2, 3), intArrayOf(4, 5, 6), intArrayOf(7, 8, 9))
            test.printMatrix(A)
            println("----------------")
            val B = test.transpose(A)
            test.printMatrix(B)

            val A2 = arrayOf(intArrayOf(1, 2, 3), intArrayOf(4, 5, 6))
            test.printMatrix(A2)
            println("----------------")
            val B2 = test.transpose(A2)
            test.printMatrix(B2)
        }
    }


    fun printMatrix(A: Array<IntArray>) {
        for (i in A.indices) {
            for (j in 0 until A[i].size) {
                print(A[i][j].toString() + " ")
            }
            println()
        }
    }
}